File: medium/LC_131_Palindrome_Partitioning/lc131.java
class Solution {
    int n;
    List<List<String>> result;
    List<String> current;
    boolean isPalindrome(String s, int l, int r){
        while(l < r){
            if(s.charAt(l) != s.charAt(r)) return false;
            l++; r--;
        }
        return true;
    }
    void backtrack(String s, int idx, List<String> current, List<List<String>> result){
        if(idx == n){
            result.add(new ArrayList(current));
            return;
        }
        for(int i = idx; i < n; i++){
            if(isPalindrome(s, idx, i)){
                current.add(s.substring(idx, i+1));
                backtrack(s, i+1, current, result);
                current.remove(current.size() - 1);
            }
        }
    }
    public List<List<String>> partition(String s) {
        n = s.length();
        result = new ArrayList();
        current = new ArrayList();
        backtrack(s, 0, current, result);
        return result;
    }
}

File: medium/LC_131_Palindrome_Partitioning/lc131.cs
public class Solution {
    int n;
    List<IList<string>> result;
    List<string> current;
    bool isPalindrome(string s, int l, int r){
        while(l < r){
            if(s[l] != s[r]) return false;
            l++; r--;
        }
        return true;
    }
    void backtrack(string s, int idx, List<string> current, List<IList<string>> result){
        if(idx == n){
            result.Add(current.ToList());
            return;
        }
        for(int i = idx; i < n; i++){
            if(isPalindrome(s, idx, i)){
                current.Add(s.Substring(idx, i-idx+1));
                backtrack(s, i+1, current, result);
                current.RemoveAt(current.Count - 1);
            }
        }
    }
    public IList<IList<string>> Partition(string s) {
        n = s.Length;
        result = new List<IList<string>>();
        current = new List<string>();
        backtrack(s, 0, current, result);
        return result;
    }
}
